js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+    if (value >> 16 === 0) return flexbuffers.BitWidth.WIDTH16;
+    if (value >> 32 === 0) return flexbuffers.BitWidth.WIDTH32;
+    return flexbuffers.BitWidth.WIDTH64;
+  }
+
+  throw `Value: ${value} is not a uint`;
+};
+
+flexbuffers.BitWidth.fromByteWidth = (value) => {
+  if (value === 1) return flexbuffers.BitWidth.WIDTH8;
+  if (value === 2) return flexbuffers.BitWidth.WIDTH16;
+  if (value === 4) return flexbuffers.BitWidth.WIDTH32;
+  if (value === 8) return flexbuffers.BitWidth.WIDTH64;
+  throw "Unexpected value " + value;
+};
+
+flexbuffers.BitWidth.paddingSize = (bufSize, scalarSize) => {
+  return (~bufSize + 1) & (scalarSize - 1);
+};
+
+flexbuffers.ValueType = {
+  NULL: 0,
+  INT: 1,
+  UINT: 2,
+  FLOAT: 3,
+  KEY: 4,
+  STRING: 5,
+  INDIRECT_INT: 6,
+  INDIRECT_UINT: 7,
+  INDIRECT_FLOAT: 8,
+  MAP: 9,
+  VECTOR: 10,
+  VECTOR_INT: 11,
+  VECTOR_UINT: 12,
+  VECTOR_FLOAT: 13,
+  VECTOR_KEY: 14,
+  VECTOR_STRING_DEPRECATED: 15,
+  VECTOR_INT2: 16,
+  VECTOR_UINT2: 17,
+  VECTOR_FLOAT2: 18,
+  VECTOR_INT3: 19,
+  VECTOR_UINT3: 20,
+  VECTOR_FLOAT3: 21,
+  VECTOR_INT4: 22,
+  VECTOR_UINT4: 23,
+  VECTOR_FLOAT4: 24,
+  BLOB: 25,
+  BOOL: 26,
+  VECTOR_BOOL: 36,
+};
+
+flexbuffers.ValueType.isInline = (value) => {
+  return value === flexbuffers.ValueType.BOOL
+    || value <= flexbuffers.ValueType.FLOAT;
+};
+
+flexbuffers.ValueType.isNumber = (value) => {
+  return value >= flexbuffers.ValueType.INT
+    && value <= flexbuffers.ValueType.FLOAT;
+};
+
+flexbuffers.ValueType.isIndirectNumber = (value) => {
+  return value >= flexbuffers.ValueType.INDIRECT_INT
+    && value <= flexbuffers.ValueType.INDIRECT_FLOAT;
+};
+
+flexbuffers.ValueType.isTypedVectorElement = (value) => {
+  return value === flexbuffers.ValueType.BOOL
+    || (value >= flexbuffers.ValueType.INT
+      && value <= flexbuffers.ValueType.STRING);
+};
+
+flexbuffers.ValueType.isTypedVector = (value) => {
+  return value === flexbuffers.ValueType.VECTOR_BOOL
+    || (value >= flexbuffers.ValueType.VECTOR_INT
+      && value <= flexbuffers.ValueType.VECTOR_STRING_DEPRECATED);
+};
+
+flexbuffers.ValueType.isFixedTypedVector = (value) => {
+  return value >= flexbuffers.ValueType.VECTOR_INT2
+    && value <= flexbuffers.ValueType.VECTOR_FLOAT4;
+};
+
+flexbuffers.ValueType.isAVector = (value) => {
+  return flexbuffers.ValueType.isTypedVector(value)
+    || flexbuffers.ValueType.isFixedTypedVector(value)
+    || value === flexbuffers.ValueType.VECTOR;
+};
+
+flexbuffers.ValueType.toTypedVector = (valueType, length) => {
+  if (length === 0) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT;
+  if (length === 2) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT2;
+  if (length === 3) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT3;
+  if (length === 4) return valueType - flexbuffers.ValueType.INT + flexbuffers.ValueType.VECTOR_INT4;
+  throw "Unexpected length " + length;
```

Comment:
Why check for validity of `length` but not `valueType`?

Reply:
We check for length mainly to select the proper value `flexbuffers.ValueType.VECTOR_INT2` | `flexbuffers.ValueType.VECTOR_INT3` | `flexbuffers.ValueType.VECTOR_INT4`. It is not really a validity check. The values behind `valueType` are chosen in a way that the formula returns a type which we expect. And yes I know it feels like magic. But it is an internal function, which is optimised for speed.